pkg/kubelet/certificate/transport.go

```diff
@@ -105,18 +106,20 @@ func addCertRotation(stopCh <-chan struct{}, period time.Duration, clientConfig
 				// the certificate has been deleted from disk or is otherwise corrupt
 				if now.After(lastCertAvailable.Add(exitAfter)) {
 					if clientCertificateManager.ServerHealthy() {
-						klog.Fatalf("It has been %s since a valid client cert was found and the server is responsive, exiting.", exitAfter)
+						klog.ErrorS(nil, "No valid client certificate is found and the server is responsive, exiting.", "lastCertificateAvailabilityTime", lastCertAvailable, "shutdownThreshold", exitAfter)
+						os.Exit(1)
 					} else {
-						klog.Errorf("It has been %s since a valid client cert was found, but the server is not responsive. A restart may be necessary to retrieve new initial credentials.", exitAfter)
+						klog.ErrorS(nil, "No valid client certificate is found but the server is not responsive. A restart may be necessary to retrieve new initial credentials.", "lastCertificateAvailabilityTime", lastCertAvailable, "shutdownThreshold", exitAfter)
 					}
 				}
 			} else {
 				// the certificate is expired
 				if now.After(curr.Leaf.NotAfter) {
 					if clientCertificateManager.ServerHealthy() {
-						klog.Fatalf("The currently active client certificate has expired and the server is responsive, exiting.")
+						klog.ErrorS(nil, "The currently active client certificate has expired and the server is responsive, exiting.")
+						os.Exit(1)
 					} else {
-						klog.Errorf("The currently active client certificate has expired, but the server is not responsive. A restart may be necessary to retrieve new initial credentials.")
+						klog.ErrorS(nil, "The currently active client certificate has expired, but the server is not responsive. A restart may be necessary to retrieve new initial credentials.")
```

Comment:
passing a nil error to ErrorS seems strange... should this do something like `klog.ErrorS(errors.New("The currently active client..."))` instead?

Reply:
Docs on calling `ErrorS` with `nil`: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#using-errors

> Calling ErrorS with nil as error is semi-acceptable if there is error condition that deserves a stack trace at this origin point. For expected errors (errors that can happen during routine operations) please consider using klog.InfoS and pass error in err key instead.

I think this is an unexpected error so this is ok?

Reply:
it looks like the klog implementation gracefully omits the err=nil from the output:

https://github.com/kubernetes/kubernetes/blob/c8fe1d9c0699f8d1fd93f47f2a62a4c83ccf5b6f/vendor/k8s.io/klog/v2/klog.go#L803-L806

but the structured logger always includes an err param, even if nil:

https://github.com/kubernetes/kubernetes/blob/c8fe1d9c0699f8d1fd93f47f2a62a4c83ccf5b6f/staging/src/k8s.io/component-base/logs/json/json.go#L171-L173

is that expected by structured log consumers?

Reply:
Yes, that's right. We just get the log message with no named error, so it's equivalent to the current behaviour.